# scMEpiTA
Single-cell Multi-layer Epigenome and Transcriptome Analysis

This project aims to specifically analyze data generated by single-cell multi-layer epigenome & transcriptome sequencing (scMEpiT-seq). There are three main modules of the project: Pre-processing including de-multiplexing, alignment, and generating cell-feature count matrix for each feature i.e., gene expression, chromatin accessibility, and other epigenomic features such as histone modifications and TF binding sites.

Pre-requisite tools:
GNU parallel (https://www.gnu.org/software/parallel/parallel_tutorial.html)
pigz (https://github.com/madler/pigz)
cutadapt (https://cutadapt.readthedocs.io/en/stable/index.html)
umi_tools (https://umi-tools.readthedocs.io/en/latest/INSTALL.html)
STAR (https://github.com/alexdobin/STAR)
featureCounts (subreads: https://subread.sourceforge.net/subread-package.html)
bedtools ()
sicer (SICER2: https://zanglab.github.io/SICER2/)
R

Installation:
cd src
make

Add the binary direction into your $PATH
The script and tools are located in scripts and bins

Usage:
Step1. Pre-porcessing.
A shell script "preprocess.sh" can go all the process in one-take action.
Step2. Core-analysis.
The cell-feature count matrix and related files generated by Step 1 can be used as input for various single-cell analyasis packages such as Seurat and Scanpy to perform normalization,  dimension reduction, clustering, cell annotation, and so on.
We provide scripts to use Seurat or Scanpy to perform these analysis on scMEpiT-seq datasets in the paper.
Step3. Uniq-analysis
scMEpiTA develops some speciall analysis tailored to the scMEpiT-seq data, including: 1 Perform single-cell ChromHMM analysis to generate chromatin states in meta-cell resolution. 
